app/browser/api/ledger.js

```diff
@@ -0,0 +1,2187 @@
+    appActions.onWalletRecovery(err, result)
+  })
+
+  return state
+}
+
+const onWalletRecovery = (state, error, result) => {
+  let existingLedgerError = ledgerState.getInfoProp(state, 'error')
+
+  if (error) {
+    // we reset ledgerInfo.error to what it was before (likely null)
+    // if ledgerInfo.error is not null, the wallet info will not display in UI
+    // logError sets ledgerInfo.error, so we must we clear it or UI will show an error
+    state = logError(error, 'recoveryWallet')
+    state = ledgerState.setInfoProp(state, 'error', existingLedgerError)
+    state = ledgerState.setRecoveryStatus(state, false)
+  } else {
+    callback(error, result)
+
+    if (balanceTimeoutId) {
+      clearTimeout(balanceTimeoutId)
+    }
+    getBalance(state)
+    state = ledgerState.setRecoveryStatus(state, true)
+  }
+
+  return state
+}
+
+const quit = (state) => {
+  quitP = true
+  state = addVisit(state, locationDefault, new Date().getTime(), null)
+
+  if (!getSetting(settings.PAYMENTS_ENABLED) && getSetting(settings.SHUTDOWN_CLEAR_HISTORY)) {
+    state = ledgerState.resetSynopsis(state, true)
+  }
+
+  return state
+}
+
+const initSynopsis = (state) => {
+  state = ledgerState.saveSynopsis(state, null, synopsis.options)
+  let value = getSetting(settings.PAYMENTS_MINIMUM_VISIT_TIME)
+  if (!value) {
+    value = 8 * 1000
+    appActions.changeSetting(settings.PAYMENTS_MINIMUM_VISIT_TIME, value)
+  }
+
+  // for earlier versions of the code...
+  if (value > 0 && value < 1000) {
+    value = value * 1000
+  }
+
+  synopsis.options.minPublisherDuration = value
+  state = ledgerState.setSynopsisOption(state, 'minPublisherDuration', value)
+
+  value = getSetting(settings.PAYMENTS_MINIMUM_VISITS)
+  if (!value) {
+    value = 1
+    appActions.changeSetting(settings.PAYMENTS_MINIMUM_VISITS, value)
+  }
+
+  if (value > 0) {
+    synopsis.options.minPublisherVisits = value
+    state = ledgerState.setSynopsisOption(state, 'minPublisherVisits', value)
+  }
+
+  if (process.env.NODE_ENV === 'test') {
+    synopsis.options.minPublisherDuration = 0
+    synopsis.options.minPublisherVisits = 0
+    state = ledgerState.setSynopsisOption(state, 'minPublisherDuration', 0)
+    state = ledgerState.setSynopsisOption(state, 'minPublisherVisits', 0)
+  } else {
+    if (process.env.LEDGER_PUBLISHER_MIN_DURATION) {
+      value = ledgerClient.prototype.numbion(process.env.LEDGER_PUBLISHER_MIN_DURATION)
+      synopsis.options.minPublisherDuration = value
+      state = ledgerState.setSynopsisOption(state, 'minPublisherDuration', value)
+    }
+    if (process.env.LEDGER_PUBLISHER_MIN_VISITS) {
+      value = ledgerClient.prototype.numbion(process.env.LEDGER_PUBLISHER_MIN_VISITS)
+      synopsis.options.minPublisherVisits = value
+      state = ledgerState.setSynopsisOption(state, 'minPublisherVisits', value)
+    }
+  }
+
+  const publishers = ledgerState.getPublishers(state)
+  for (let item of publishers) {
```

Comment:
for loop should be much faster performance-wise

Reply:
we are for looping Immutable.Map, which I am not sure how can I do it with index based for loop. I use this approach vs `publishers.forEach()` because it should be faster.